src/plots/cartesian/axes.js

```diff
@@ -506,102 +507,186 @@ axes.autoBin = function(data, ax, nbins, is2d) {
     }
 
     return {
-        start: ax.c2r(binstart),
-        end: ax.c2r(binend),
-        size: dummyax.dtick
+        start: ax.c2r(binStart),
+        end: ax.c2r(binEnd),
+        size: dummyAx.dtick
     };
 };
 
 
+function autoShiftNumericBins(binStart, data, ax, dataMin, dataMax) {
+    var edgecount = 0,
+        midcount = 0,
+        intcount = 0,
+        blankCount = 0;
+
+    function nearEdge(v) {
+        // is a value within 1% of a bin edge?
+        return (1 + (v - binStart) * 100 / ax.dtick) % 100 < 2;
+    }
+
+    for(var i = 0; i < data.length; i++) {
+        if(data[i] % 1 === 0) intcount++;
+        else if(!isNumeric(data[i])) blankCount++;
+
+        if(nearEdge(data[i])) edgecount++;
+        if(nearEdge(data[i] + ax.dtick / 2)) midcount++;
+    }
+    var dataCount = data.length - blankCount;
+
+    if(intcount === dataCount && ax.type !== 'date') {
+        // all integers: if bin size is <1, it's because
+        // that was specifically requested (large nbins)
+        // so respect that... but center the bins containing
+        // integers on those integers
+        if(ax.dtick < 1) {
+            binStart = dataMin - 0.5 * ax.dtick;
+        }
+        // otherwise start half an integer down regardless of
+        // the bin size, just enough to clear up endpoint
+        // ambiguity about which integers are in which bins.
+        else binStart -= 0.5;
+    }
+    else if(midcount < dataCount * 0.1) {
+        if(edgecount > dataCount * 0.3 ||
+                nearEdge(dataMin) || nearEdge(dataMax)) {
+            // lots of points at the edge, not many in the middle
+            // shift half a bin
+            var binshift = ax.dtick / 2;
+            binStart += (binStart + binshift < dataMin) ? binshift : -binshift;
+        }
+    }
+    return binStart;
+}
+
+
+function autoShiftMonthBins(binStart, data, dtick, dataMin) {
+    var exactYears = 0,
+        exactMonths = 0,
+        exactDays = 0,
+        blankCount = 0,
+        dataCount,
+        di,
+        d,
+        year,
+        month;
+
+    for(var i = 0; i < data.length; i++) {
+        di = data[i];
+        if(!isNumeric(di)) {
+            blankCount ++;
+            continue;
+        }
+        d = new Date(di),
+        year = d.getUTCFullYear();
+        if(di === Date.UTC(year, 0, 1)) {
+            exactYears ++;
+        }
+        else {
+            month = d.getUTCMonth();
+            if(di === Date.UTC(year, month, 1)) {
+                exactMonths ++;
+            }
+            else if(di === Date.UTC(year, month, d.getUTCDate())) {
+                exactDays ++;
+            }
+        }
+    }
+
+    dataCount = data.length - blankCount;
+
+    // include bigger exact dates in the smaller ones
+    exactMonths += exactYears;
+    exactDays += exactMonths;
+
+    // unmber of data points that needs to be an exact value
+    // to shift that increment to (near) the bin center
+    var threshold = 0.8 * dataCount;
```

Comment:
non- :no_entry: It would be nice to put these auto-algo threshold fractions in a `constants.js` file somewhere. It's not obvious where that file should be though.